### DISCUSSION TITLE
Twitch Plugin Client-Id

### DISCUSSION THREAD

**Discussion by dennismamedov** [streamlink/streamlink#4320] [Q&A] [answered]
Hey!

I'm planning to write an application with this project as a dependency. 
As I was looking through the source code I found a hard-coded Client-Id which is being used for all requests to Twitch's GQL API. I am very curious where this client-id comes from. Is this a Client-Id someone owns, or is this a guaranteed future-proof Client-Id?

The Client-Id is mentioned at [this](https://github.com/streamlink/streamlink/blob/711418045debdc37fb9ea6f42fc737872b8eb342/src/streamlink/plugins/twitch.py#L217) line.

Thanks for any help!

**Comment by bastimeyer** [comment, marked as answer]
This client-ID is from Twitch's own website which uses their private GraphQL API.

Streamlink's Twitch plugin was previously using their public kraken API with its own client-ID (check the plugin's git history). This client-ID had been stolen and misused a couple of times by other projects in the past instead of them registering their own, at least to what I know, so thanks for asking.

The kraken API will be shut down at the end of February this year though and Twitch's new public API called helix unfortunately requires user authentication for every request, which is not great if all that Streamlink needs from this API is stream metadata like user name, stream title, stream category/game, etc.

Acquiring a streaming access token is also only possible on GQL nowadays, which was previously done on their private REST API, similar to kraken. This private REST API has been shut down last year and it previously introduced breaking changes where we had to completely remove authentication for Twitch due to OAuth tokens being bound to client-IDs and thus being incompatible between different APIs / namespaces. That's why the decision was made to completely ignore their public APIs, which are limited in use for Streamlink anyway, and completely use their private GQL API for all requests.

As said, the GQL API client-ID is from Twitch's own website, and it's not a secret. It is however not meant to be used by third parties, but there's no other choice, unfortunately, and other projects need to use it, too.

**Reply by dennismamedov** [reply]
Thank you for your great response, Sebastian. This clears a lot of things up for me, and I hope for others encountering this discussion in the future!
